share/man/man4/esp.4
.\"	$NetBSD$
.\"
.\" Copyright (c) 1997 Eric S. Hvozda <hvozda@netcom.com>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed by Eric S. Hvozda.
.\" 4. The name of Eric S. Hvozda may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	$NetBSD$
.\"
.Dd December 3, 2001
.Dt ESP 4
.Os
.Sh NAME
.Nm esp
.Nd NCR 53C9x, Emulex ESP406, and Qlogic FAS408 SCSI driver
.Sh SYNOPSIS
.Ss ISA bus
.Cd "esp0 at isa? port 0x230 irq ?"
.Ss PCMCIA
.Cd "esp* at pcmcia? function ?"
.Ss MCA
.Cd "esp* at mca? slot ?"
.Ss mac68k
.Cd "esp0 at obio?"
.Cd "esp1 at obio?"
.Ss macppc
.Cd "esp0 at obio0 flags 0x00ff"
.Ss sun3x
.Cd "esp0 at obio0 addr 0x66000000 ipl 2 flags 0xff0f"
.Ss sparc
.Cd "dma0 at obio0 addr 0xfa001000 level 4" Pq Sun 4/300
.Cd "esp0 at obio0 addr 0xfa000000 level 4" Pq Sun 4/300
.Pp
.Cd "dma0 at sbus0 slot ? offset ?" Pq sun4c and sun4m
.Cd "esp0 at sbus0 slot ? offset ?" Pq sun4c
.Cd "esp0 at dma0" Pq sun4m
.Pp
.Cd "dma* at sbus? slot ? offset ?" Pq Sbus
.Cd "esp* at sbus? slot ? offset ?" Pq SBus, older PROMs
.Cd "esp* at dma?" Pq SBus
.Pp
.Cd "scsibus* at esp?"
.Sh DESCRIPTION
The
.Nm
driver provides support for the
.Tn NCR
53C90, 53C94 and 53C96;
.Tn Emulex
ESP100, ESP100A, ESP200 and ESP406; and
.Tn Qlogic
FAS216 and FAS408
.Tn SCSI
controller chips found in a wide variety of systems and peripheral boards.
This includes the
.Tn Qlogic
.Tn ISA
and
.Tn VLB
.Tn SCSI
host adapters, and the
.Tn Sun
Fast
.Tn SCSI
buffered
.Tn Ethernet
for
.Tn Sbus
.Pq FSBE/S, X1053A, Sun part # 501-2015 .
.Pp
For
.Tn Qlogic
.Tn PCI
.Tn SCSI
host adapters, use the
.Xr isp 4
device.
.Sh CONFIGURATION
The
.Nm
driver supports the following
.Sy flags
for use in
.Xr config 1
files:
.Pp
.Bl -tag -compact -width "bits 16-23:"
.It bits 0-7 :
disable disconnect/reselect for the corresponding
.Tn SCSI
target
.It bits 8-15 :
disable synchronous negotiation for the corresponding
.Tn SCSI
target
.It bits 16-23 :
disable tagged queuing for the corresponding
.Tn SCSI
target
.El
.Pp
.Qq Target
is synonymous with
.Tn SCSI
ID number.
.Pp
Note that
.Tn SCSI
tape drives should be allowed to perform disconnect/reselect or performance
will suffer.
.Sh SEE ALSO
.Xr cd 4 ,
.Xr ch 4 ,
.Xr intro 4 ,
.Xr le 4 ,
.Xr mca 4 ,
.Xr pcmcia 4 ,
.Xr scsi 4 ,
.Xr sd 4 ,
.Xr ss 4 ,
.Xr st 4 ,
.Xr uk 4
.\" .Sh BUGS
.\" Neither
.\" .Tn DMA
.\" or synchronous data transfers are currently supported.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

sys/arch/sun3/dev/esp.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 1997 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Jeremy Cooper and Gordon W. Ross
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * "Front end" glue for the ncr53c9x chip, formerly known as the
 * Emulex SCSI Processor (ESP) which is what we actually have.
 */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD$");

#include <sys/types.h>
#include <sys/param.h>
#include <sys/systm.h>
#include <sys/kernel.h>
#include <sys/errno.h>
#include <sys/device.h>
#include <sys/buf.h>

#include <dev/scsipi/scsi_all.h>
#include <dev/scsipi/scsipi_all.h>
#include <dev/scsipi/scsiconf.h>
#include <dev/scsipi/scsi_message.h>

#include <machine/autoconf.h>
#include <machine/bus.h>

#include <dev/ic/ncr53c9xreg.h>
#include <dev/ic/ncr53c9xvar.h>

#include <sun3/dev/dmareg.h>
#include <sun3/dev/dmavar.h>

#define	ESP_REG_SIZE	(12*4)

struct esp_softc {
	struct ncr53c9x_softc sc_ncr53c9x;	/* glue to MI code */
	bus_space_tag_t sc_bst;			/* bus space tag */
	bus_space_handle_t sc_bsh;		/* bus space handle */
	struct dma_softc *sc_dma;		/* pointer to my dma */
};

static int	espmatch(device_t, cfdata_t, void *);
static void	espattach(device_t, device_t, void *);

CFATTACH_DECL_NEW(esp, sizeof(struct esp_softc),
    espmatch, espattach, NULL, NULL);

/*
 * Functions and the switch for the MI code.
 */
static uint8_t	esp_read_reg(struct ncr53c9x_softc *, int);
static void	esp_write_reg(struct ncr53c9x_softc *, int, uint8_t);
static int	esp_dma_isintr(struct ncr53c9x_softc *);
static void	esp_dma_reset(struct ncr53c9x_softc *);
static int	esp_dma_intr(struct ncr53c9x_softc *);
static int	esp_dma_setup(struct ncr53c9x_softc *, uint8_t **, size_t *,
		    int, size_t *);
static void	esp_dma_go(struct ncr53c9x_softc *);
static void	esp_dma_stop(struct ncr53c9x_softc *);
static int	esp_dma_isactive(struct ncr53c9x_softc *);

static struct ncr53c9x_glue esp_glue = {
	esp_read_reg,
	esp_write_reg,
	esp_dma_isintr,
	esp_dma_reset,
	esp_dma_intr,
	esp_dma_setup,
	esp_dma_go,
	esp_dma_stop,
	esp_dma_isactive,
	NULL,			/* gl_clear_latched_intr */
};

static int 
espmatch(device_t parent, struct cfdata *cf, void *aux)
{
	struct confargs *ca = aux;

	/*
	 * Check for the esp registers.
	 */
	if (bus_peek(ca->ca_bustype,
	    ca->ca_paddr + (NCR_STAT * 4), 1) == -1)
		return 0;

	/* If default ipl, fill it in. */
	if (ca->ca_intpri == -1)
		ca->ca_intpri = 2;

	return 1;
}

static void 
espattach(device_t parent, device_t self, void *aux)
{
	struct esp_softc *esc = device_private(self);
	struct ncr53c9x_softc *sc = &esc->sc_ncr53c9x;
	struct confargs *ca = aux;

	/*
	 * Set up glue for MI code early; we use some of it here.
	 */
	sc->sc_dev = self;
	sc->sc_glue = &esp_glue;

	/*
	 * Map the ESP registers.
	 */
	esc->sc_bst = ca->ca_bustag;
	if (bus_space_map(esc->sc_bst, ca->ca_paddr, ESP_REG_SIZE, 0,
	    &esc->sc_bsh) != 0) {
		aprint_error(": can't map register\n");
		return;
	}

	/* Other settings */
	sc->sc_id = 7;
	sc->sc_freq = 20;	/* The 3/80 esp runs at 20 MHz */

	/*
	 * Hook up the DMA driver.
	 */
	esc->sc_dma = espdmafind(device_unit(self));
	esc->sc_dma->sc_client = sc; /* Point back to us */

	/*
	 * XXX More of this should be in ncr53c9x_attach(), but
	 * XXX should we really poke around the chip that much in
	 * XXX the MI code?  Think about this more...
	 */

	/*
	 * It is necessary to try to load the 2nd config register here,
	 * to find out what rev the esp chip is, else the ncr53c9x_reset
	 * will not set up the defaults correctly.
	 */
	sc->sc_cfg1 = sc->sc_id | NCRCFG1_PARENB;
	sc->sc_cfg2 = NCRCFG2_SCSI2 | NCRCFG2_RPE;
	sc->sc_cfg3 = NCRCFG3_CDB;
	NCR_WRITE_REG(sc, NCR_CFG2, sc->sc_cfg2);

	if ((NCR_READ_REG(sc, NCR_CFG2) & ~NCRCFG2_RSVD) !=
	    (NCRCFG2_SCSI2 | NCRCFG2_RPE)) {
		sc->sc_rev = NCR_VARIANT_ESP100;
	} else {
		sc->sc_cfg2 = NCRCFG2_SCSI2;
		NCR_WRITE_REG(sc, NCR_CFG2, sc->sc_cfg2);
		sc->sc_cfg3 = 0;
		NCR_WRITE_REG(sc, NCR_CFG3, sc->sc_cfg3);
		sc->sc_cfg3 = (NCRCFG3_CDB | NCRCFG3_FCLK);
		NCR_WRITE_REG(sc, NCR_CFG3, sc->sc_cfg3);
		if (NCR_READ_REG(sc, NCR_CFG3) !=
		    (NCRCFG3_CDB | NCRCFG3_FCLK)) {
			sc->sc_rev = NCR_VARIANT_ESP100A;
		} else {
			/* NCRCFG2_FE enables > 64K transfers */
			sc->sc_cfg2 |= NCRCFG2_FE;
			sc->sc_cfg3 = 0;
			NCR_WRITE_REG(sc, NCR_CFG3, sc->sc_cfg3);
			sc->sc_rev = NCR_VARIANT_ESP200;
		}
	}

	/*
	 * XXX minsync and maxxfer _should_ be set up in MI code,
	 * XXX but it appears to have some dependency on what sort
	 * XXX of DMA we're hooked up to, etc.
	 */

	/*
	 * This is the value used to start sync negotiations
	 * Note that the NCR register "SYNCTP" is programmed
	 * in "clocks per byte", and has a minimum value of 4.
	 * The SCSI period used in negotiation is one-fourth
	 * of the time (in nanoseconds) needed to transfer one byte.
	 * Since the chip's clock is given in MHz, we have the following
	 * formula: 4 * period = (1000 / freq) * 4
	 */
	sc->sc_minsync = 1000 / sc->sc_freq;

	/*
	 * Alas, we must now modify the value a bit, because it's
	 * only valid when can switch on FASTCLK and FASTSCSI bits
	 * in config register 3...
	 */
	switch (sc->sc_rev) {
	case NCR_VARIANT_ESP100:
		sc->sc_maxxfer = 64 * 1024;
		sc->sc_minsync = 0;	/* No synch on old chip? */
		break;

	case NCR_VARIANT_ESP100A:
		sc->sc_maxxfer = 64 * 1024;
		/* Min clocks/byte is 5 */
		sc->sc_minsync = ncr53c9x_cpb2stp(sc, 5);
		break;

	case NCR_VARIANT_ESP200:
		sc->sc_maxxfer = 16 * 1024 * 1024;
		/* XXX - do actually set FAST* bits */
		break;
	}

	/* and the interuppts */
	isr_add_autovect(ncr53c9x_intr, sc, ca->ca_intpri);
	evcnt_attach_dynamic(&sc->sc_intrcnt, EVCNT_TYPE_INTR, NULL,
	    device_xname(self), "intr");

	/* Do the common parts of attachment. */
	sc->sc_adapter.adapt_minphys = minphys;
	sc->sc_adapter.adapt_request = ncr53c9x_scsipi_request;
	ncr53c9x_attach(sc);

	/* Turn on target selection using the `dma' method */
	sc->sc_features |= NCR_F_DMASELECT;
}


/*
 * Glue functions.
 */

uint8_t
esp_read_reg(struct ncr53c9x_softc *sc, int reg)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	return bus_space_read_1(esc->sc_bst, esc->sc_bsh, reg * 4);
}

void
esp_write_reg(struct ncr53c9x_softc *sc, int reg, uint8_t val)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	bus_space_write_1(esc->sc_bst, esc->sc_bsh, reg * 4, val);
}

int 
esp_dma_isintr(struct ncr53c9x_softc *sc)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	return DMA_ISINTR(esc->sc_dma);
}

void 
esp_dma_reset(struct ncr53c9x_softc *sc)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	dma_reset(esc->sc_dma);
}

int 
esp_dma_intr(struct ncr53c9x_softc *sc)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	return espdmaintr(esc->sc_dma);
}

int 
esp_dma_setup(struct ncr53c9x_softc *sc, uint8_t **addr, size_t *len,
    int datain, size_t *dmasize)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	return dma_setup(esc->sc_dma, addr, len, datain, dmasize);
}

void 
esp_dma_go(struct ncr53c9x_softc *sc)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	DMA_GO(esc->sc_dma);
}

void 
esp_dma_stop(struct ncr53c9x_softc *sc)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	DMA_STOP(esc->sc_dma);
}

int 
esp_dma_isactive(struct ncr53c9x_softc *sc)
{
	struct esp_softc *esc = (struct esp_softc *)sc;

	return DMA_ISACTIVE(esc->sc_dma);
}
